samples/java-protobuf-customer-registry-kafka-quickstart/src/it/java/customer/api/CustomerActionWithKafkaIntegrationTest.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.*;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.junit.Assert.assertTrue;
+import static org.testcontainers.shaded.org.awaitility.Awaitility.await;
+
+// This class was initially generated based on the .proto definition by Kalix tooling.
+//
+// As long as this file exists it will not be overwritten: you can maintain it yourself,
+// or delete it so it is regenerated as needed.
+
+// Example of an integration test calling our service via the Kalix proxy
+// IMPORTANT: this tests depends on an external kafka instance. Make sure to have it running
+// Run all test classes ending with "IntegrationTest" using `mvn verify -Pit`
+public class CustomerActionWithKafkaIntegrationTest {
+
+  /**
+   * The test kit starts both the service container and the Kalix proxy.
+   */
+  @ClassRule
+  public static final KalixTestKitResource testKit =
+    new KalixTestKitResource(Main.createKalix(),
+        KalixTestKit.Settings.DEFAULT.withEventingSupport(KalixTestKit.Settings.EventingSupport.KAFKA));
+
+  /**
+   * Use the generated gRPC client to call the service through the Kalix proxy.
+   */
+  private final CustomerService client;
+
+
+  public CustomerActionWithKafkaIntegrationTest() {
+    client = testKit.getGrpcClient(CustomerService.class);
+  }
+
+  @Test
+  public void createAndPublish() throws Exception {
+    var id = UUID.randomUUID().toString();
+    var customer = buildCustomer(id, "Johanna", "foo@example.com", "Porto", "Long Road");
+    createCustomer(customer);
+
+    //change customer name
+    var completeName = "Johanna Doe";
+    client.changeName(CustomerApi.ChangeNameRequest.newBuilder()
+            .setCustomerId(id)
+            .setNewName(completeName)
+            .build())
+        .toCompletableFuture()
+        .get(5, SECONDS);
+
+
+    // Example Kafka consumer configuration
+    Properties props = new Properties();
+    props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "0.0.0.0:9093");
+    props.put(ConsumerConfig.GROUP_ID_CONFIG, "it-test-" + System.currentTimeMillis()); // using new consumer group every run
+    props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+    props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+    props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
```

Comment:
Messy that it has to be repeated here, but maybe enough. Probably not _that_ common that users will think they need Kafka integration tests?

Reply:
Yeah, my rational is that users will only need one of this when they are interested in testing something specific related with Kafka and its clients. e.g. make sure that certain messages go to certain partitions. In these cases, I think it's reasonable to expose users to these details.

For general business related tests, testkit mocked topic is the way to go.